composition-js/src/merging/merge.ts

```diff
@@ -2572,6 +2584,116 @@ class Merger {
       // Because we rename all root type in subgraphs to their default names, we shouldn't ever have incompatibilities here.
       assert(!isIncompatible, () => `Should not have incompatible root type for ${rootKind}`);
     }
+    this.addJoinDirectiveDirectives(sources, dest);
+  }
+
+  private normalizeSpecIdentityURL(identityURL: string): string {
+    const url = new URL(identityURL);
+    const pathParts = url.pathname.split('/');
+    if (/v\d+\.\d+/.test(pathParts[pathParts.length - 1])) {
+      // Remove the specific version to obtain the canonical URL.
+      pathParts.pop();
+      url.pathname = pathParts.join('/');
+    }
+    return url.toString();
+  }
+
+  private shouldUseJoinDirectiveForURL(url: string | undefined): boolean {
+    return Boolean(
+      url &&
+      this.joinDirectiveIdentityURLs.has(this.normalizeSpecIdentityURL(url))
+    );
+  }
+
+  private computeMapFromImportedNamesToIdentityURLs(
+    schema: Schema,
+  ): Map<string, string> {
+    // For each @link directive on the schema definition, store its normalized
+    // identity url in a Map, reachable from all its imported names.
+    const map = new Map<string, string>();
+
+    for (const linkDirective of schema.schemaDefinition.appliedDirectivesOf<LinkDirectiveArgs>('link')) {
+      const { url, import: imps } = linkDirective.arguments();
+      const identityURL = this.normalizeSpecIdentityURL(url);
+      if (imps) {
+        for (const imp of imps) {
+          if (typeof imp === 'string') {
+            map.set(imp, identityURL);
+          } else {
+            map.set(imp.as ?? imp.name, identityURL);
+          }
+        }
+      }
+    }
+    return map;
+  }
+
+  // This method gets called at various points during the merge to allow
+  // subgraph directive applications to be reflected (unapplied) in the
+  // supergraph, using the @join__directive(graphs,name,args) directive.
+  private addJoinDirectiveDirectives(
+    sources: (SchemaElement<any, any> | undefined)[],
+    dest: SchemaElement<any, any>,
+  ) {
+    const joinsByDirectiveName: {
+      [directiveName: string]: Array<{
+        graphs: string[];
+        args: Record<string, any>;
+      }>
+    } = Object.create(null);
+
+    for (const [idx, source] of sources.entries()) {
+      if (!source) continue;
+      const graph = this.joinSpecName(idx);
+
+      // We compute this map only once per subgraph, as it takes time
+      // proportional to the size of the schema.
+      const linkImportIdentityURLMap =
+        this.computeMapFromImportedNamesToIdentityURLs(source.schema());
+
+      for (const directive of source.appliedDirectives) {
+        const isRelevantLink =
```

Comment:
Is there a reason we want to add @link? I'll probably get to it later on, but I would avoid this if we can.

Reply:
I think it's important to preserve the `@link` associated with importing any of the directives we're wrapping with `@join__directive`, because the `@link` tells us whether/how the imported directives have been renamed or prefixed. Without that information, there'd be no way to tell which of the original `@source*` directives is intended when the supergraph SDL contains `@join__directive(graphs: [...], name: "sourceRenamed", args: {...})`.

An alternative would be to disallow renaming the `@source*` directives in subgraph schemas, but that's a surprising gotcha to impose, and not consistent with how `@link` typically works.

Including `@link` is something @sachindshinde recommended, and I agree it's worthwhile.